### DISCUSSION TITLE
Mattermost 9.11

### DISCUSSION THREAD

**Discussion by hnykda** [coollabsio/coolify#3087] [Service Template Requests]
I managed to run Mattermost, here is the config as Docker Compose type of deployment:

```
services:
  postgres:
    image: 'postgres:${POSTGRES_IMAGE_TAG}'
    restart: '${RESTART_POLICY}'
    security_opt:
      - 'no-new-privileges:true'
    pids_limit: 100
    read_only: true
    tmpfs:
      - /tmp
      - /var/run/postgresql
    volumes:
      - '${POSTGRES_DATA_PATH}:/var/lib/postgresql/data'
    environment:
      - TZ
      - POSTGRES_USER
      - POSTGRES_PASSWORD
      - POSTGRES_DB
  mattermost:
    depends_on:
      - postgres
    image: 'mattermost/${MATTERMOST_IMAGE}:${MATTERMOST_IMAGE_TAG}'
    restart: '${RESTART_POLICY}'
    security_opt:
      - 'no-new-privileges:true'
    pids_limit: 200
    read_only: '${MATTERMOST_CONTAINER_READONLY}'
    tmpfs:
      - /tmp
    volumes:
      - './config:/mattermost/config:rw'
      - './data:/mattermost/data:rw'
      - './logs:/mattermost/logs:rw'
      - './plugins:/mattermost/plugins:rw'
      - './client/plugins:/mattermost/client/plugins:rw'
      - './bleve-indexes:/mattermost/bleve-indexes:rw'
    environment:
      - TZ
      - MM_SQLSETTINGS_DRIVERNAME
      - MM_SQLSETTINGS_DATASOURCE
      - MM_BLEVESETTINGS_INDEXDIR
      - MM_SERVICESETTINGS_SITEURL
    expose:
      - 8065
```

Env file:
```
DOMAIN=<you-domain>
MATTERMOST_CONTAINER_READONLY=false
MATTERMOST_IMAGE=mattermost-enterprise-edition
MATTERMOST_IMAGE_TAG=9.11
MM_BLEVESETTINGS_INDEXDIR=/mattermost/bleve-indexes
# probably not needed
MM_SERVICESETTINGS_SITEURL=https://<your-domain>
MM_SQLSETTINGS_DATASOURCE=postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@postgres:5432/${POSTGRES_DB}?sslmode=disable&connect_timeout=10
MM_SQLSETTINGS_DRIVERNAME=postgres
POSTGRES_DATA_PATH=./volumes/db/var/lib/postgresql/data
POSTGRES_DB=mattermost
POSTGRES_IMAGE_TAG=13-alpine
POSTGRES_PASSWORD=mmuser_password
POSTGRES_USER=mmuser
RESTART_POLICY=unless-stopped
TZ=Europe/Prague
```

After saving the file, I had to go to the coolify directory of the service, e.g. `/data/coolify/services/isgow40goosk0osksc8wco8s` on my server, and run: 

```
sudo chown -R 2000:2000 bleve-indexes/ client/ config/ data/ logs plugins
```

then, after setting my domain (including https! - `https://<my-domain>`) in `Domains` under the mattermost service, it works.

More detailed guide [below](https://github.com/coollabsio/coolify/discussions/3087#discussioncomment-10490690). I also used Mattermost official [docker docs](https://docs.mattermost.com/install/install-docker.html#deploy-mattermost-on-docker-for-production-use):
![image](https://github.com/user-attachments/assets/c86e6795-f1dd-4e2f-b1e2-92900aa4df78)

You don't need nginx and certificates from that guide, you are going to use Coolify's traefik automagic instead.

If you want push notifications on mobile devices, you also want to allow TPNS in mattermost admin console:
![image](https://github.com/user-attachments/assets/0009b579-e687-4571-9a93-1f5333ed5b07)

## Optional
For backups, I have in my docker-compose file this service:
```
  dbbackups:
    image: 'kartoza/pg-backup:15-3.3'
    volumes:
      - './db-backups:/backups'
    environment:
      - POSTGRES_HOST=postgres
      - POSTGRES_USER=mmuser
      - POSTGRES_PASS=mmuser_password
      - REMOVE_BEFORE=7
      - DUMPPREFIX=mattermost
    restart: on-failure
    depends_on:
      postgres:
        condition: service_healthy
```

which  makes daily dumps in that data folder (which I backup regularly).

**Comment by MaxAryus** [comment]
I get the following error when starting the service…

Starting containers.
time="2024-08-29T17:21:11Z" level=warning msg="The \"POSTGRES_USER\" variable is not set. Defaulting to a blank string."
time="2024-08-29T17:21:11Z" level=warning msg="The \"POSTGRES_PASSWORD\" variable is not set. Defaulting to a blank string."
time="2024-08-29T17:21:11Z" level=warning msg="The \"POSTGRES_DB\" variable is not set. Defaulting to a blank string."

Edit:

I guess the issue comes form the Mattermost container. In the logs it says:

` Error: failed to load configuration: could not create config file: open /mattermost/config/config.json: permission denied`

**Reply by hnykda** [reply]
Eh, I think I saw that too, but it's working nevertheless. You can try to hardcode them in the docker-compose directly.

**Reply by MaxAryus** [reply]
I did so but then the Error: `Error: failed to load configuration: could not create config file: open /mattermost/config/config.json: permission denied` shows up.

I tried opening the url but it says `404 page not found`

**Reply by MaxAryus** [reply]
In generell. I experienced that all .env values are not working I needed to add them directly to the docker compose file.

**Reply by hnykda** [reply]
OK, weird, I didn't have that issue. Did you run that `chown` command?

**Comment by RowanAldean** [comment]
Can someone just give me a ELI5 step-by-step for getting this sorted on my Hetzner VPS? First time using Coolify

**Reply by hnykda** [reply]
Well, what have you tried so far? If your coolify setup works (e.g. domain etc.), you should be fine to:

1. go to projects -> resources -> select/create new project -> click New under resources
2. when creating new resource, click on Docker Compose under "Docker based"
3. into the "Docker Compose" text area that is on the screen insert the docker-compose I shared above
4. set domain, click save
![image](https://github.com/user-attachments/assets/7a4b2fcf-50f2-436d-b4ea-87a52e5cd212)
5. set env vars from above (click developer view), then click save
![image](https://github.com/user-attachments/assets/1db20143-d376-4e4b-a1bd-041cb63575cb)
6. you shouldn't IMHO need to set up volumes as they should get fetched from the docker-compose, but the page should look like this: 
![image](https://github.com/user-attachments/assets/c2b067e3-1d90-4521-a685-04586dc26bfa)
7. then you have to do the slightly weird task. You have to ssh to your VPS, and as a root go to the data directory of this service. For me, it was `/data/coolify/services/isgow40goosk0osksc8wco8s` (it's the same ID that is in the URL when setting this up), and run that `sudo chown -R 2000:2000 bleve-indexes/ client/ config/ data/ logs plugins`. I found out that this step is needed in in the [Mattermost docker deployment use](https://docs.mattermost.com/install/install-docker.html#deploy-mattermost-on-docker-for-production-use).
8. then go to coolify and click `Deploy` button

**Reply by RowanAldean** [reply]
Can't reach mattermost via 8065. Github didn't refresh so only just seeing this comment now when I refreshed my page! 

Going to continue troubleshooting. Doesn't help that my domain DNS just points to the VPS. I don't know if I can subdomain at random like that :) i.e i'd need an A record for `mattermost.<domain>.com`

**edit:** bruh, just reading your reply about adding the domain via the UI. Trying this now.

**Comment by RowanAldean** [comment]
Got this functional but having trouble with TLS (using 443) for Mattermost. Anyone got any ideas?

**Reply by RowanAldean** [reply]
Ok, the last line you've added explains this - I don't need to worry as using Coolify Traefik magic. Going to assume that gives me encrypted comms (I suspected out-of-the-box was sufficient).

**Reply by hnykda** [reply]
Yeah, exactly. Don't ask me how it works, I just pointed `*.mydomain` to my VPS and let coolify traefik magic do it's thing.

**Reply by RowanAldean** [reply]
I'm doubtful on this actually being encyrpted 😂 I did setup the certs and CORS stuff though so now browser seems happy and things are functional. Just gotta move over our small team now!

**Reply by hnykda** [reply]
I am pretty confident it is getting encrypted, actually, thanks to letsencrytp and traefik etc. 😄  Good luck!

**Comment by RowanAldean** [comment]
Getting the github bot working with this stuff is ridiculous. Wasted so many hours with this today that i'm wondering if Coolify was even "cool".

Ridiculously overcomplicated. Getting 307s following docs or "Text not defined" if trying to just use `Incoming Webhooks`

**Comment by benjamin-steurer** [comment]
Nice tutorial:

For me it was applications instead of services.

Unfortunately somehow the env files won't work. I get this when starting it:
time="2024-09-06T21:06:50+02:00" level=warning msg="The \"POSTGRES_USER\" variable is not set. Defaulting to a blank string."

**Reply by peaklabs-dev** [reply]
Please make sure you are updated to the latest version, as there was a fix for "POSTGRES_USER" in it (just add the compose file again).

**Reply by benjamin-steurer** [reply]
@peaklabs-dev coolify is one v4.0.0-beta.330. 

Still not working. Wonder if its related to the fact, that the volumes are in "applications" instead of services.

`time="2024-09-07T13:50:38+02:00" level=warning msg="The \"POSTGRES_USER\" variable is not set. Defaulting to a blank string."
time="2024-09-07T13:50:38+02:00" level=warning msg="The \"POSTGRES_DB\" variable is not set. Defaulting to a blank string."`

**Reply by peaklabs-dev** [reply]
Could you please open a detailed issue with reproduction steps. Thanks.

**Reply by PsychoDude** [reply]
I had the exact same issue, got the latest version to work (10.0.0) thanks to Darren on discord. You are indeed correct, I had to chown the folders inside applications also. This is the Compose to get you up and running:

```
services:
  postgres:
    image: 'postgres:16-alpine'
    restart: unless-stopped
    security_opt:
      - 'no-new-privileges:true'
    pids_limit: 100
    read_only: true
    tmpfs:
      - /tmp
      - /var/run/postgresql
    volumes:
      - 'postgres:/var/lib/postgresql/data'
    environment:
      - TZ=UTC
      - 'POSTGRES_USER=${SERVICE_USER_POSTGRES:-postgres}'
      - 'POSTGRES_PASSWORD=${SERVICE_PASSWORD_POSTGRES}'
      - 'POSTGRES_DB=${DB_POSTGRES:-postgres}'
  mattermost:
    depends_on:
      - postgres
    image: 'mattermost/mattermost-enterprise-edition:10.0.0'
    restart: unless-stopped
    security_opt:
      - 'no-new-privileges:true'
    pids_limit: 200
    read_only: false
    tmpfs:
      - /tmp
    volumes:
      - './config:/mattermost/config:rw'
      - './data:/mattermost/data:rw'
      - './logs:/mattermost/logs:rw'
      - './plugins:/mattermost/plugins:rw'
      - './client/plugins:/mattermost/client/plugins:rw'
      - './bleve-indexes:/mattermost/bleve-indexes:rw'
    environment:
      - SERVICE_FQDN_MATTERMOST_8065
      - TZ=UTC
      - MM_SQLSETTINGS_DRIVERNAME=postgres
      - 'MM_SQLSETTINGS_DATASOURCE=postgres://${SERVICE_USER_POSTGRES}:${SERVICE_PASSWORD_POSTGRES}@postgres:5432/${DB_POSTGRES:-postgres}?sslmode=disable&connect_timeout=10'
      - MM_BLEVESETTINGS_INDEXDIR=/mattermost/bleve-indexes
      - 'MM_SERVICESETTINGS_SITEURL=${SERVICE_FQDN_MATTERMOST}'
    expose:
      - 8065
```

No need to generate any env variables, they get auto-generated. It assigns a wildcard url if you have one set up for your server. 

Note that the mattermost container does take a few minutes to boot up.

* You need to start a fresh service for this to work

**Comment by robots4life** [comment]
Running this on a CX22 | x86 | 40 GB | instance, take note, Mattermost does not work on ARM instances !

**Mattermost 10.8.1**

Environment

Just paste into Environment Variables > Developer View

```text
# Domain of service
DOMAIN=domain.com

# Container settings
## Timezone inside the containers. The value needs to be in the form 'Europe/Berlin'.
## A list of these tz database names can be looked up at Wikipedia
## https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
TZ=UTC
RESTART_POLICY=unless-stopped

# Postgres settings
## Documentation for this image and available settings can be found on hub.docker.com
## https://hub.docker.com/_/postgres
## Please keep in mind this will create a superuser and it's recommended to use a less privileged
## user to connect to the database.
## A guide on how to change the database user to a nonsuperuser can be found in docs/creation-of-nonsuperuser.md
POSTGRES_IMAGE_TAG=13-alpine
POSTGRES_DATA_PATH=./volumes/db/var/lib/postgresql/data

POSTGRES_USER=user
POSTGRES_PASSWORD=password
POSTGRES_DB=mattermost

# Nginx
## The nginx container will use a configuration found at the NGINX_MATTERMOST_CONFIG. The config aims
## to be secure and uses a catch-all server vhost which will work out-of-the-box. For additional settings
## or changes ones can edit it or provide another config. Important note: inside the container, nginx sources
## every config file inside */etc/nginx/conf.d* ending with a *.conf* file extension.

## Inside the container the uid and gid is 101. The folder owner can be set with
## `sudo chown -R 101:101 ./nginx` if needed.
## Note that this repository requires nginx version 1.25.1 or later
## NGINX_IMAGE_TAG=alpine

## The folder containing server blocks and any additional config to nginx.conf
## NGINX_CONFIG_PATH=./nginx/conf.d
## NGINX_DHPARAMS_FILE=./nginx/dhparams4096.pem

## CERT_PATH=./volumes/web/cert/cert.pem
## KEY_PATH=./volumes/web/cert/key-no-password.pem
#GITLAB_PKI_CHAIN_PATH=<path_to_your_gitlab_pki>/pki_chain.pem
#CERT_PATH=./certs/etc/letsencrypt/live/${DOMAIN}/fullchain.pem
#KEY_PATH=./certs/etc/letsencrypt/live/${DOMAIN}/privkey.pem

## Exposed ports to the host. Inside the container 80, 443 and 8443 will be used
## HTTPS_PORT=443
## HTTP_PORT=80
## CALLS_PORT=8443

# Mattermost settings
## Inside the container the uid and gid is 2000. The folder owner can be set with
## `sudo chown -R 2000:2000 ./volumes/app/mattermost`.
MATTERMOST_CONFIG_PATH=./volumes/app/mattermost/config
MATTERMOST_DATA_PATH=./volumes/app/mattermost/data
MATTERMOST_LOGS_PATH=./volumes/app/mattermost/logs
MATTERMOST_PLUGINS_PATH=./volumes/app/mattermost/plugins
MATTERMOST_CLIENT_PLUGINS_PATH=./volumes/app/mattermost/client/plugins
MATTERMOST_BLEVE_INDEXES_PATH=./volumes/app/mattermost/bleve-indexes

## Bleve index (inside the container)
MM_BLEVESETTINGS_INDEXDIR=/mattermost/bleve-indexes

## This will be 'mattermost-enterprise-edition' or 'mattermost-team-edition' based on the version of Mattermost you're installing.
MATTERMOST_IMAGE=mattermost-team-edition
## Update the image tag if you want to upgrade your Mattermost version. You may also upgrade to the latest one. The example is based on the latest Mattermost ESR version.
MATTERMOST_IMAGE_TAG=10.8.1

## Make Mattermost container readonly. This interferes with the regeneration of root.html inside the container. Only use
## it if you know what you're doing.
## See https://github.com/mattermost/docker/issues/18
MATTERMOST_CONTAINER_READONLY=false

## The app port is only relevant for using Mattermost without the nginx container as reverse proxy. This is not meant
## to be used with the internal HTTP server exposed but rather in case one wants to host several services on one host
## or for using it behind another existing reverse proxy.
APP_PORT=8065

## Configuration settings for Mattermost. Documentation on the variables and the settings itself can be found at
## https://docs.mattermost.com/administration/config-settings.html
## Keep in mind that variables set here will take precedence over the same setting in config.json. This includes
## the system console as well and settings set with env variables will be greyed out.

## Below one can find necessary settings to spin up the Mattermost container
MM_SQLSETTINGS_DRIVERNAME=postgres
MM_SQLSETTINGS_DATASOURCE=postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@postgres:5432/${POSTGRES_DB}?sslmode=disable&connect_timeout=10

## Example settings (any additional setting added here also needs to be introduced in the docker-compose.yml)
MM_SERVICESETTINGS_SITEURL=https://${DOMAIN}
```

Docker Compose

```yml
# https://docs.docker.com/compose/environment-variables/
services:
  postgres:
    image: postgres:${POSTGRES_IMAGE_TAG}
    restart: ${RESTART_POLICY}
    security_opt:
      - no-new-privileges:true
    pids_limit: 100
    read_only: true
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U ${POSTGRES_USER} -d ${POSTGRES_DB}"]
      interval: 10s
      timeout: 5s
      retries: 5
      start_period: 10s
    tmpfs:
      - /tmp
      - /var/run/postgresql
    volumes:
      - ${POSTGRES_DATA_PATH}:/var/lib/postgresql/data
    environment:
      # timezone inside container
      - TZ

      # necessary Postgres options/variables
      - POSTGRES_USER
      - POSTGRES_PASSWORD
      - POSTGRES_DB

  mattermost:
    depends_on:
      postgres:
        condition: service_healthy
    image: mattermost/${MATTERMOST_IMAGE}:${MATTERMOST_IMAGE_TAG}
    restart: ${RESTART_POLICY}
    security_opt:
      - no-new-privileges:true
    pids_limit: 200
    read_only: ${MATTERMOST_CONTAINER_READONLY}
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:8065/api/v4/system/ping"]
      interval: 30s
      timeout: 10s
      retries: 3
      start_period: 30s
    tmpfs:
      - /tmp
    volumes:
      - ${MATTERMOST_CONFIG_PATH}:/mattermost/config:rw
      - ${MATTERMOST_DATA_PATH}:/mattermost/data:rw
      - ${MATTERMOST_LOGS_PATH}:/mattermost/logs:rw
      - ${MATTERMOST_PLUGINS_PATH}:/mattermost/plugins:rw
      - ${MATTERMOST_CLIENT_PLUGINS_PATH}:/mattermost/client/plugins:rw
      - ${MATTERMOST_BLEVE_INDEXES_PATH}:/mattermost/bleve-indexes:rw
      # When you want to use SSO with GitLab, you have to add the cert pki chain of GitLab inside Alpine
      # to avoid Token request failed: certificate signed by unknown authority
      # (link: https://github.com/mattermost/mattermost-server/issues/13059 and https://github.com/mattermost/docker/issues/34)
      # - ${GITLAB_PKI_CHAIN_PATH}:/etc/ssl/certs/pki_chain.pem:ro
    environment:
      # timezone inside container
      - TZ

      # necessary Mattermost options/variables (see env.example)
      - MM_SQLSETTINGS_DRIVERNAME
      - MM_SQLSETTINGS_DATASOURCE

      # necessary for bleve
      - MM_BLEVESETTINGS_INDEXDIR

      # additional settings
      - MM_SERVICESETTINGS_SITEURL
    expose:
      - 8065
# If you use rolling image tags and feel lucky watchtower can automatically pull new images and
# instantiate containers from it. https://containrrr.dev/watchtower/
# Please keep in mind watchtower will have access on the docker socket. This can be a security risk.
#
#  watchtower:
#    container_name: watchtower
#    image: containrrr/watchtower:latest
#    restart: unless-stopped
#    volumes:
#      - /var/run/docker.sock:/var/run/docker.sock
```

No need to do anything with any folder permissions on the VPS or any containers any more, works flawless without that step.

Make sure on your VPS you have port 8065 open for incoming connections.

In the Mattermost Service Settings you put the domain, also here no need to put the port after it, just the domain.

![image](https://github.com/user-attachments/assets/70f1b961-af8c-49ff-8b65-d2133e406ccb)
